src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,401 @@
+        auto pattern2_unsigned = cast(Int(32, 0), cast(Int(32, 0), wild_u8x) * wild_i32x);
+
+        if (!(expr_match(pattern2, reduce->value, matches) || expr_match(pattern2_unsigned, reduce->value, matches))) {
+            return {};
+        }
+    } else {
+        auto pattern2 = cast(Float(32, 0), cast(Float(32, 0), wild_bf16x) * wild_f32x);
+
+        if (!expr_match(pattern2, reduce->value, matches)) {
+            return {};
+        }
+    }
+
+    const auto *lhs_load = matches[0].as<Load>();
+    const auto *rhs_broadcast = matches[1].as<Broadcast>();
+    if (!lhs_load || !rhs_broadcast) {
+        return {};
+    }
+    const auto *rhs_cast = rhs_broadcast->value.as<Cast>();
+    if (rhs_cast) {
+        if (op_type == AMXOpType::Int8) {
+            if (!(rhs_cast->value.type().element_of() == Int(8) || rhs_cast->value.type().element_of() == UInt(8))) {
+                user_assert(false) << "Expected rhs cast of i8/u8";
+            }
+        } else  // AMXOpType::Bf16
+        {
+            user_assert(rhs_cast->value.type().element_of() == BFloat(16)) << "Expected rhs cast of bf16";
+        }
+    } else {
+        return {};
+    }
+
+    const auto *rhs_load = rhs_cast->value.as<Load>();
+    if (!rhs_load) {
+        return {};
+    }
+
+    const auto lhs_tile = is_3d_tile_index(lhs_load->index);
+    const auto rhs_tile = is_2d_tile_index(rhs_load->index);
+    if (!lhs_tile.result || !rhs_tile.result) {
+        return {};
+    }
+
+    const int tile_x = lhs_tile.extent[0];
+    const int tile_y = lhs_tile.extent[1];
+    const int tile_r = lhs_tile.extent[2];
+    const int factor = reduce->value.type().lanes() / reduce->type.lanes();
+    if (op->index.type().lanes() != tile_x * tile_y ||
+        factor != tile_r ||
+        tile_y != rhs_tile.extent[0] ||
+        tile_r != rhs_tile.extent[1]) {
+        return {};
+    }
+
+#if LLVM_VERSION < 130
+    user_assert(op_type != AMXOpType::Bf16 &&
+                lhs_load->type.is_int() && rhs_cast->value.type().is_int())
+        << "LLVM 13 or above is required for unsigned or float AMX instructions";
+#endif
+
+    // {rows, colbytes, var, index}
+    auto lhs_var = Variable::make(Handle(), lhs_load->name);
+    const auto &lhs_load_type = lhs_load->type;
+    int element_width = lhs_load_type.bytes();
+    auto lhs_type = lhs_load_type.with_lanes(1024 / element_width);
+    auto lhs = Call::make(lhs_type, "tile_load", {tile_x, tile_r * element_width, lhs_var, lhs_tile.base * element_width, lhs_tile.stride[0] * element_width}, Call::Intrinsic);
+
+    auto rhs_var = Variable::make(Handle(), rhs_load->name);
+    const auto &rhs_load_type = rhs_load->type;
+    auto rhs_type = rhs_load_type.with_lanes(1024 / element_width);
+    auto rhs = Call::make(rhs_type, "tile_load", {1, tile_y * tile_r * element_width, rhs_var, rhs_tile.base * element_width, rhs_tile.stride[0] * tile_y * element_width}, Call::Intrinsic);
+
+    auto res_type = [&]() {
```

Comment:
This will cause the `res_type` var to always be fully initialized from the lambdas return value, rather than first partially initializing with the default constructor and assigning afterwards. I can change it if it doesn't fit the Halide style.

Reply:
Since the same lambda appears in several places, I think making a helper function to convert these types would be better.

Reply:
I've added the following helper function `Type amx_op_type_result_type(AMXOpType)` to convert each case to the proper `Type`.